[GeoMechanicsApplycation] Settlement API

rfaasse opened the pull request:
**📝 Description**
This PR ties the components for the Settlement workflow together and fixes several issues. This means stages 1-3 of the test case (see also the added cpp test in this PR) are run successfully with the expected results.

**🆕 Changelog**
- Added the implementation of the strategy wrapper
- Added the relevant code to the dgeosettlement.h/.cpp files. 
- Added an integration test for the settlement workflow
- Extended and refactored the write_output.h/.cpp which is responsible for writing output to the GID res format
- Imported the LinearSolversApplication as a direct dependency in the settlement workflow (this also means the GeoMechanicsApplication library now depends through cmake on the LinearSolversApplication)

WPK4FEM committed:
WPK first attempt time loop executor

WPK4FEM committed:
also complementary to WPK first attempt time loop executor

WPK4FEM committed:
Removed yet unused things. Added first unit test: TimeLoopReturnsPerformedStatesAfterRunning. Removed template argument from TimeLoopExecutor class.

WPK4FEM committed:
Added a StrategyWrapper with an always converging solver strategy for testing

WPK4FEM committed:
Added constructor with convergence argument to DummySolverStrategy. Added test for running and terminating a loop with a non-converging solver strategy.

WPK4FEM committed:
Give time loop executor a TimeStepEndState to start from. Keep track of the time passed. Added a test on end time after N steps.

WPK4FEM committed:
Use the time step executor. ( only one available so default constructed ). Passed solver strategy to time step executor.

WPK4FEM committed:
Keep track of used number of cycles. Added tests using a test time incrementor with a fixed number of cycles per step. Added RunCycle and RunCycleLoop to the time loop executor.

avdg81 committed:
Removed a few unused dummy classes for testing

Also added the `explicit` keyword to a single-argument constructor.

avdg81 committed:
Test the time loop executor using a prescribed time incrementor

Given a prescribed time incrementor that keeps four time increments and
a solver strategy that always converges, we expect to get four converged
time step end states, where each one just required a single cycle. Also
the time values are validated for each time step end state.

Furthermore added a helper function to build a converged end state.

avdg81 committed:
Added the strategy wrapper as a member of the time loop executor

Now that the strategy wrapper is a member of the time loop executor, we
can set the end time and the time increment in their "global" variables.
Internally, this is done via the model part. The reason we set these
variables explicitly is that processes may rely on having this
information available.

avdg81 committed:
When running the time loop, increment the step number

Extended the API of class `StrategyWrapper` such that we can increment
the step number and get the step number upon request.

WPK4FEM committed:
Added test for and implementation of CloneTimeStep, in the strategy wrapper needed for the time loop executor.

rfaasse committed:
Merge branch 'master' into settlement_test

WPK4FEM committed:
Added method and tests for RestorePositionsAndDOFVectorToStartOfStep for restoring to the initial situation of a step to enable cycling of the step.

rfaasse committed:
First implementation of the strategy wrapper implementation

rfaasse committed:
Added unit tests for all non-trivial functionality of the wrapper

rfaasse committed:
Fix pipeline build

WPK4FEM committed:
SaveTotalDisplacementFieldAtStartOfStage and AccumulateTotalDisplacementField implemented and unit tests provided.

WPK4FEM committed:
OutputProcess implementation and unit test added to time_loop_executor.

WPK4FEM committed:
FinalizeSolutionStep should be called outside the cycle loop. This changes a test for the TimeStepExecutor and there is a new one for the TimeLoopExecutor.

WPK4FEM committed:
Small things from review by Richard adapted.

WPK4FEM committed:
Pass on process observables from TimeLoopExecutor to TimeStepExecutor.

WPK4FEM committed:
Less code repetition in test_time_loop_executor

WPK4FEM committed:
interface class for TimeLoopExecutor as requested in review by Richard.

WPK4FEM committed:
merge from master

rfaasse committed:
Added all the new strategy wrapper functions

WPK4FEM committed:
forgotten interface.h

WPK4FEM committed:
Code smell remarks from SonarCloud almost all resolved

rfaasse committed:
Added UTs for Getters, setters and Save and Accumulate

rfaasse committed:
Added output process to strategy wrapper

rfaasse committed:
Added functionality for restore positions and dof vector

WPK4FEM committed:
Implemented suggestions from review by Anne.

WPK4FEM committed:
Last? SonarCloud code smell fixed

rfaasse committed:
Added unit test Restoring DoFs

rfaasse committed:
Merge branch 'geo/time-step-loop' into geo/strategy_wrapper_implementation

rfaasse committed:
Merge time loop executor work into feature branch

rfaasse committed:
Added test

rfaasse committed:
Merge branch 'settlement_test' into geo/strategy_wrapper_implementation

rfaasse committed:
Added integration tests

avdg81 committed:
Several small improvements

- No longer track the log file that is created by running the unit test.
- Removed an unused `#include`.
- Use a `std::filesystem::path` object to represent the working
  directory. Also renamed this variable to comply with the Kratos Style
  Guide.
- Corrected a mistake in the names of the project parameter files (the
  word "stage" was missing).
- Open the log file that is created by the unit test only once. And use
  write mode rather than append mode.

avdg81 committed:
Extracted copying of nodal solution step values

When the solving strategy wrapper needs to restore the nodal positions
and the nodal solution step values, it explicitly checks for each nodal
solution step variable whether the model part actually has it. If yes,
then it copies for each node the value at the given source index to the
given destination index.

Split an existing unit test into two new tests. The first one verifies
whether the combination of displacements and water pressures can be
restored. The second one verifies whether the combination of
displacements and rotations can be restored.

avdg81 committed:
Refactored and improved some unit tests

- Extracted a helper function that creates a dummy model part.
- The helper function that creates a solving strategy wrapper now no
  longer creates a dummy model part. Rather it receives one through its
  parameter list.
- Gave some test values a name.
- Corrected some usages of `KRATOS_EXPECT_EQ` to
  `KRATOS_EXPECT_DOUBLE_EQ`.

avdg81 committed:
Refactored and improved some unit tests

- The helper function that creates a solving strategy wrapper with
  default process info entries now no longer creates a dummy model part.
  Rather it receives one through its parameter list.
- Eliminated a few unneeded local variables.
- The test for incrementing the step number now increments and verifies
  the step number twice.
- Corrected a few variable names to comply with the style guide.

avdg81 committed:
Added the adaptive time incrementor to the settlement workflow

- Added several helper functions to extract the parameters for the
  adaptive time incrementor from the project parameters.
- The time loop executor of the settlement workflow now receives an
  adaptive time incrementor.
- The factory that creates the settlement workflow now uses a valid time
  loop executor.

carloslubbers committed:
Merge branch 'master' into geo/strategy_wrapper_implementation

# Conflicts:
#	applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.cpp

carloslubbers committed:
Add logging when new time step is executed

avdg81 committed:
Added a working solving strategy wrapper to the settlement workflow

- Create a solving strategy wrapper instance and pass it to the time
  loop executor of the settlement workflow.
- The solving strategy factory now accepts "newton_raphson" as strategy
  type rather than "newton_raphson_strategy" (which is the name returned
  by that strategy). This is in line with the Python workflow of the
  settlement calculation.

rfaasse committed:
Added linear solvers application link to settlement and geomechanics

avdg81 committed:
The settlement workflow now tries to run the time loop executor

For now, we pass a dummy time step end state to the `Run` member
function.

Note that the unit test of the settlement workflow now ends with a
crash!

avdg81 committed:
Corrected a mistake when creating the linear solvers application

- Corrected the type of data member `mpLinearSolverApp`. It used to be
  `KratosGeoMechanicsApplication::Pointer`, whereas
  `KratosLinearSolversApplication::Pointer` was intended.
- Also corrected the statement that creates this smart pointer.
- Use the plural form (linear solvers) in the name of the data member,
  to be consistent with the name of the application.

avdg81 committed:
Add degrees of freedom to after reading the `.mdpa` file

This is in line with what the D-GeoFlow workflow does and it seems to
resolve a crash. (Carlos had discovered the same thing.)

avdg81 committed:
Print lots of debug messages and fixed a few minor issues

- Implemented the `Info` member functions of classes
  `ApplyK0ProcedureProcess`, `ApplyScalarConstraintTableProcess`,
  `ApplyVectorConstraintTableProcess`, and
  `BackwardEulerQuasistaticUPwScheme`.
- Explicitly set the number of threads to be used by the settlement
  workflow to 1 (in line with what is done by the geoflow workflow).
- Included the fix for the model parts used by the process creators for
  the settlement workflow.
- Properly initialize the processes used by the settlement workflow.
- Added a new model part that is referenced by the K0 procedure.
- Added the K0 procedure to the list of processes for the first stage of
  the example settlement workflow.
- Added a few checks for "empty" objects when creating a solver
  strategy.
- For testing purposes, we have removed the conditions entirely from the
  `.mdpa` file of the example settlement analysis. Furthermore, we have
  removed all the submodel parts that are not used during the first
  stage of the example settlement analysis.

rfaasse committed:
Added missing gauss parameters to writing the output. The settlement workflow test now passes for stage 1

avdg81 committed:
Restored the conditions and several submodel parts

For testing purposes, those had been temporarily removed. Since we now
handle the model parts given to the processes in a better way, we should
try to restore those.

rfaasse committed:
Made fix for option to omit hydraulic head at nodes and removed log file for test to fix pipeline. First three stages now complete without crashing

rfaasse committed:
Added python res files

avdg81 committed:
Create a computational model part

In line with the approach followed by the Python scripts, we need to
have a computational (sub)model part that keeps "active" parts of the
mesh/model, including the nodes, elements, and conditions of the main
model part. It also shares the process info with the main model part.
This computational model part is then passed to the factory that creates
solving strategies.

When the solving strategy wrapper attempts to clone a time step, we now
either pass its model part (if it is not a submodel part) or the root
model part of its (sub)model part.

Furthermore, we added a few more print statements for debugging.

rfaasse committed:
Removed logging from core library

avdg81 committed:
Fixed the times written to the GiD output files

Also accepted changed results for the D-GeoFlow "unit tests".

rfaasse committed:
reverting core files to master state

avdg81 committed:
Fixed several code smells reported by SonarQube

- Replaced a deprecated function call by its successor.
- Changed the types of two variables to a reference-to-const.
- Changed the type of a parameter to a reference-to-const.

rfaasse committed:
Fix GCC build

rfaasse committed:
Removed unnecessary logging

rfaasse committed:
Extracted duplicate code to fix build

rfaasse committed:
Merge branch 'master' into geo/strategy_wrapper_implementation

rfaasse committed:
Small fix after merge

WPK4FEM committed:
Introduced the spatial tables of Marjan into the settlement test model

WPK4FEM committed:
Merge remote-tracking branch 'refs/remotes/origin/geo/strategy_wrapper_implementation' into geo/strategy_wrapper_implementation

rfaasse committed:
Reduced code smells

WPK4FEM committed:
GeoElementUtilities::CalculateNodalHydraulicHeadFromWaterPressures now gets its size for the returned vector from the number of nodes of the element.

WPK4FEM committed:
Merge remote-tracking branch 'refs/remotes/origin/geo/strategy_wrapper_implementation' into geo/strategy_wrapper_implementation

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_dgeosettlement.cpp:
```diff
@@ -83,8 +85,9 @@ KRATOS_TEST_CASE_IN_SUITE(RunStageMakesRelevantCallsOnce, KratosGeoMechanicsFast
-    RunStage(settlement);
+    const auto exit_status = RunStage(settlement);
 
+    KRATOS_EXPECT_EQ(exit_status, 1); // Ask Richard why we should expect an error state here
```
@avdg81 Saw your question in the comment and this is probably because the parameter_json_settings (see line 84) don't have all info needed for the solving strategy factory, so I expect that will throw now. We can either leave it with an explanation, create another interface + stub for the factory, or provide all necessary inputs

avdg81 replied:
I would suggest that we add a brief explanation for now. Let's revisit these tests later on.

rfaasse replied:
Done

rfaasse committed:
Remove redundant print statement

avdg81 committed:
Merge branch 'geo/strategy_wrapper_implementation' of github.com:KratosMultiphysics/Kratos into geo/strategy_wrapper_implementation

avdg81 committed:
Simplified the writing of integration point (and nodal) results

We no longer need to cast and test a variable to determine which
function overload we need for writing the variable's results. We now
build up a map of integration point (and nodal) result writers, which in
turn are created by two factory function templates. Adding a new result
to be output is a matter of adding a single line.

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.cpp:
```diff
@@ -24,6 +24,69 @@
+using LinearSolverType = LinearSolver<SparseSpaceType, DenseSpaceType>;
+using SolvingStrategyFactoryType = SolvingStrategyFactory<SparseSpaceType, DenseSpaceType, LinearSolverType>;
+
+double GetStartTimeFrom(const Parameters& rProjectParameters)
```
Let's keep in mind that we have these functions and move it to some utilities if we have need for them somewhere else

avdg81 replied:
Agreed.

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.cpp:
```diff
@@ -131,9 +206,24 @@ int KratosGeoSettlement::RunStage(const std::filesystem::path&            rWorki
+            mpTimeLoopExecutor->SetTimeIncrementor(MakeTimeIncrementor(project_parameters));
+            mpTimeLoopExecutor->SetSolverStrategyWrapper(MakeStrategyWrapper(project_parameters,
+                                                                             rWorkingDirectory));
+            // For now, pass a dummy state. THIS PROBABLY NEEDS TO BE REFINED AT SOME POINT!
```
Do we want to keep this, or fix it in this PR?

avdg81 replied:
I would say keep it for now, and let's try to address it later this week.

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.cpp:
```diff
@@ -235,6 +322,93 @@ std::vector<std::shared_ptr<Process>> KratosGeoSettlement::GetProcesses(const Pa
+std::shared_ptr<StrategyWrapper> KratosGeoSettlement::MakeStrategyWrapper(const Parameters&            rProjectParameters,
+                                                                          const std::filesystem::path& rWorkingDirectory)
+{
+    // For now, we can create solving strategy wrappers only
```
Do we need this comment?

avdg81 replied:
I think it's helpful to see that at present we only create one type of strategy wrapper. But I have reordered these lines a bit to make it more logical.

rfaasse committed:
Added copying properties

avdg81 committed:
Removed a redundant blank line

rfaasse committed:
removed redundant log file

rfaasse commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_solver.py:
```diff
@@ -216,6 +216,7 @@ def KeepAdvancingSolutionLoop(self, end_time):
         return current_time_corrected < end_time
 
     def Initialize(self):
+        KratosMultiphysics.Logger.PrintInfo("GeoMechanicalSolver", "Initialize()")
```
In my opinion we can remove these logging statements

avdg81 replied:
Agreed. I have removed them.

rfaasse committed:
Cleaning up

rfaasse committed:
Merge branch 'geo/strategy_wrapper_implementation' of https://github.com/KratosMultiphysics/Kratos into geo/strategy_wrapper_implementation

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_time_stepping.cpp:
```diff
@@ -79,6 +79,9 @@ class DummyStrategyWrapper : public StrategyWrapper
     [[nodiscard]] unsigned int NumberOfSolverStrategyInitializeCalls()             const {return mSolverStrategyInitializeCalls;}
     [[nodiscard]] unsigned int NumberOfSolverStrategyInitializeSolutionStepCalls() const {return mSolverStrategyInitializeSolutionStepCalls;}
+
+    ~DummyStrategyWrapper() override = default;
```
Is this necessary?

avdg81 replied:
No, I don't think so. So I will remove it.

rfaasse committed:
Moved settlement workflow test to integration suite of cpp tests

avdg81 committed:
Reordered a few lines to make the flow more logical

rfaasse committed:
Removing SQ issues

rfaasse committed:
Merge remote-tracking branch 'origin/geo/strategy_wrapper_implementation' into geo/strategy_wrapper_implementation

avdg81 commented on applications/GeoMechanicsApplication/custom_workflows/solving_strategy_wrapper.hpp:
```diff
@@ -0,0 +1,184 @@
+    {
+        if (mResetDisplacements)
+        {
+            for (const auto& node : mrModelPart.Nodes())
+            {
+                mOldTotalDisplacements.emplace_back(node.GetSolutionStepValue(TOTAL_DISPLACEMENT));
+            }
```
When this member function is called more than once, it will _append_ the current total displacement field, rather than replace it (which would be the expected behavior, I guess). So we may want to add `mOldTotalDisplacements.clear()` prior to emplacing the current total displacement field.

rfaasse replied:
Done

avdg81 committed:
Two minor enhancements

1. Prefer to use the pre-increment operator rather than the
   post-increment operator.
2. Made a local variable `const`.

carloslubbers commented on applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.cpp:
```diff
@@ -44,6 +107,14 @@ KratosGeoSettlement::KratosGeoSettlement(std::unique_ptr<InputUtility> pInputUti
+        mKernel.ImportApplication(mpLinearSolversApp);
+    }
+
+    ParallelUtilities::SetNumThreads(1);
```
Can this be decided on by the value in the project parameters file? Jonathan suggested we should use this parameter to test the effect of multi-threading in the future

rfaasse replied:
@carloslubbers Good suggestion! We'll definitely keep it in mind, but we would like to do that as part of the next PR (since we have some issues for stage 4 we also need to tackle). That way, we can get this part to master and you can already test stage 1-3 in DSettlement

carloslubbers replied:
@rfaasse  No problem, good to have it in a future PR

avdg81 committed:
Removed several log statements

avdg81 committed:
Removed a redundant statement

A node that has been created by a model part does not need to be added
explicitly to that same model part later on.

avdg81 committed:
Removed a redundant destructor

rfaasse committed:
Added test for saving total replacements twice and other minor fixes

rfaasse committed:
Fix Codacy complaint

rfaasse committed:
Fix UT

rfaasse committed:
Only run stage 1-3 in the workflow test, since stage 4 has known issues

rfaasse committed:
Merge branch 'master' into geo/strategy_wrapper_implementation

rfaasse committed:
Possible fix for pipeline test failure

rfaasse committed:
Another attempt at fixing the pipeline test

rfaasse committed:
Skipping workflow test temporarily due to discussed pipeline issues

avdg81 reviewed (dismissed):
Good enough for now.

rfaasse committed:
Disabling workflow test temporarily due to discussed pipeline issues

avdg81 reviewed (approved):
Good enough for now. Let's try to continue with our work :-)
